test/clib/test_ctfunc.cpp

```diff
@@ -4,101 +4,540 @@
 
 #include "cantera/core.h"
 #include "cantera/clib/ctfunc.h"
-#include "cantera/numerics/Func1.h"
+#include "cantera/numerics/Func1Factory.h"
 
 using namespace Cantera;
 
+string reportType(int fcn)
+{
+    int buflen = func_type(fcn, 0, 0);
+    char* buf = new char[buflen];
+    func_type(fcn, buflen, buf);
+    string funcName = buf;
+    delete[] buf;
+    return funcName;
+}
+
+
+TEST(ctfunc, functor)
+{
+    int fcn = func_new_basic("functor", 0.);
+    ASSERT_GE(fcn, 0);
+    ASSERT_EQ(reportType(fcn), "functor");
+    ASSERT_EQ(func_derivative(fcn), -1); // not implemented
+    ASSERT_EQ(func_del(999), -1);
+    ASSERT_EQ(func_del(fcn), 0);
+    ASSERT_EQ(func_derivative(fcn), -1); // function no longer exists
+}
+
+TEST(ctfunc, invalid)
+{
+    // exceptions return -1
+    ASSERT_EQ(func_new_basic("spam", 0.), -1);
+    ASSERT_EQ(func_new_advanced("eggs", 1, 0, NULL), -1);
+}
+
 TEST(ctfunc, sin)
 {
     double omega = 2.;
-    int fcn = func_new(SinFuncType, 0, 1, &omega);
+    int fcn = func_new_basic("sin", omega);
     ASSERT_GE(fcn, 0);
+    ASSERT_EQ(reportType(fcn), "sin");
     EXPECT_DOUBLE_EQ(func_value(fcn, 0.), 0.);
     EXPECT_DOUBLE_EQ(func_value(fcn, 0.5), sin(omega * 0.5));
+
+    auto functor = newFunc1("sin", omega);
+    EXPECT_DOUBLE_EQ(functor->eval(0.5), func_value(fcn, 0.5));
+
+    int dfcn = func_derivative(fcn);
+    EXPECT_DOUBLE_EQ(func_value(dfcn, 0.), omega);
+    EXPECT_DOUBLE_EQ(func_value(dfcn, 0.5), omega * cos(omega * 0.5));
+
+    auto dfunctor = functor->derivative3();
+    EXPECT_DOUBLE_EQ(dfunctor->eval(0.5), func_value(dfcn, 0.5));
```

Comment:
What does testing the `Func1` object directly this way do that isn't covered by testing via the clib calls?

Reply:
Indeed there isn't much new information here. However, in absence of a pure C++ test suite, I'd rather have a little redundancy here.